src/node/eviction_impl.h

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_NODE_EVICTION_IMPL_H
+#define BITCOIN_NODE_EVICTION_IMPL_H
+
+#include <node/connection_types.h>
+#include <net_permissions.h>
+#include <sync.h>
+
+#include <cstdint>
+#include <map>
+#include <optional>
+#include <vector>
+
+typedef int64_t NodeId;
+
+struct NodeEvictionCandidate {
+    NodeId id;
+    std::chrono::seconds m_connected;
+    std::chrono::microseconds m_min_ping_time;
+    std::chrono::seconds m_last_block_time;
+    std::chrono::seconds m_last_tx_time;
+    bool fRelevantServices;
+    bool m_relay_txs;
+    bool fBloomFilter;
+    uint64_t nKeyedNetGroup;
+    bool prefer_evict;
+    bool m_is_local;
+    Network m_network;
+    bool m_noban;
+    ConnectionType m_conn_type;
+};
+
+[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates);
```

Comment:
Is there any reason why this and `ProtectEvictionCandidatesByRatio()` aren't members of `EvictionManagerImpl`? I'm not saying they have to be, but I'm interested in the design choice.

Reply:
This is something that @theuni  originally also had on his branch and i just went with the same design. To me this seems good because some of the tests purely test these functions without needing the eviction manager and adding them as members would require more test refactoring.

Reply:
I think that test refactoring would probably be a good thing, since it would ensure that the unit test is not testing a state that EvictionManager can't get into in the product code. In general, that seems like a good model - unit tests use the public methods to update the state of the object being tested, but optionally also have access to const test-only methods to peek inside the state of the object if it's difficult to expose that state in other ways.

Certainly not required in this PR. Could be done in a follow-up (or not at all).